### ISSUE TITLE
[FEATURE] PCI-E passthrough (and GPU support)

### ISSUE THREAD

**Issue by A-vesalius** [harvester/harvester#992]
**Is your feature request related to a problem? Please describe.**
VM cannot utilize the full network capability of baremetal hardware, especially 10gbe and above as well as DPDK with harvester.

**Describe the solution you'd like**
Leverage Multus which has the ability to use pci-e passthrough, IOMMU and VT-d, devices and SR-IOV VFs for kubernetes networks. Once these VF's aresetup on baremetal then harvestor GUI should allow the user to hand them out to VMs in an otherwise transparent way equivalent to physical nics. 

**Describe alternatives you've considered**
https://github.com/k8snetworkplumbingwg/sriov-network-device-plugin#quick-start

https://docs.mellanox.com/pages/releaseview.action?pageId=25145021

**Additional context**

**Comment by tlehman** [comment]
Just to clarify what SR-IOV gives you, I want to make a simple diagram that motivates why this is such a high priority:

When a packet arrives at the NIC, there's one core that is interrupted and handles that interrupt. Then the destination VM's core is interrupted and stops, and the hypervisor copies the packet to the VM's memory space, then the VM core continues, and so does the core that handles the NIC's interrupts.

## Without SR-IOV:

```mermaid
sequenceDiagram
Packet ->> NIC: 2 packets arrive
NIC ->> Hypervisor: copy packet 1
Hypervisor ->> VM3: copy packet 1
VM3 ->> VM3: continues
NIC ->> Hypervisor: copy packet 2
Hypervisor ->> VM6: copy packet 2
VM6 ->> VM6: continues
````

## With SR-IOV
```mermaid
sequenceDiagram
Packet ->> NIC: 2 packets arrive
NIC ->> VF3: queued
NIC ->> VF6: queued
VF3 ->> VM3: DMA to VM
VF6 ->> VM6: DMA to VM
```

(DMA is short for Direct Memory Access)

In summary, SR-IOV cuts the hypervisor out of the pathway from the NIC to the VM, dramatically increasing the throughput in your virtualized environment.

**Comment by w13915984028** [comment]
connect with https://github.com/harvester/harvester/issues/2236
SRIOV may also be represented as a kind of network to reuse existing features

**Comment by gusraggio** [comment]
Rancher is using this chart to configure this, https://github.com/rancher/charts/pull/1211#pullrequestreview-668492770

**Comment by tlehman** [comment]
> Rancher is using this chart to configure this, [rancher/charts#1211 (review)](https://github.com/rancher/charts/pull/1211#pullrequestreview-668492770)

Interesting approach, the [numVfs is modeled as a policy](https://github.com/rancher/charts/pull/1211/files#diff-936fc461e3d600fd20b7fabb04eb148973be077ec07354d78a39486753565560R100). My branch used a Harvester settings object to store the number of vfs, but I'll borrow the Rancher way, to use the prior art.

**Comment by WuJun2016** [comment]
Hi， @mantis-toboggan-md  Can you take a look at the issue ?   https://github.com/tlehman/harvester/blob/sriov-enhancement/enhancements/20220614-sriov-configuration.md#requirements  thanks

**Comment by mantis-toboggan-md** [comment]
Yeah @WuJun2016 I'll sync up with Tobi next week to discuss

**Comment by tlehman** [comment]
@A-vesalius Could you describe your intended use of this feature? What sort of workloads should be make sure we support?

**Comment by A-vesalius** [comment]
My particular purpose would be to pass through network adapters to allow for full speed use by VM’s specifically direct hardware offloading of network task to the nic as opposed to being forced to use the vm/host cpu when using virtualized or paravirtualized nics. 25, 40 and 100GbE nics for instance if used in a harvestor vm to external physical storage network or a virtualized router VM would be limited with only virtualized/paravirtualized interfaces such as VirtIO. 

secondarily while not my intended purpose it would be great if this was built to also allowed for passing through of GPU's to VM's under harvestor. Many uses there as well.

**Comment by tlehman** [comment]
Got it, that's what I am working on in this ticket, enabling SR-IOV for baremetal Harvester cases, allowing the VMs to make full use of the NICs. @A-vesalius  Can you give more detail about your likely workloads? Are you thinking about having a single VF per VM? Or Multiple?

Also, GPU passthrough is going to be covered in #1029

**Comment by A-vesalius** [comment]
@tlehman Mainly thinking of single VF's to a VM, but have plans for multiple as well, with LACP and redundant connections to the VM in a segregated vlan or even the host network. Although in those instances I was more thinking of distributing VF's from port A, B, C, or D of a 4 port PCI card to a single VM and not as much sending VF1 and VF2 from only port A to a single VM. If possible giving people that option would always be better.

**Comment by tlehman** [comment]
@A-vesalius what network hardware are you using?

**Comment by mircea-pavel-anton** [comment]
I would like to leave a note on another use-case for PCIe passthrough, which is passing storage controllers to VMs. For example, my storage server is a virtualized instance of TrueNAS and I am passing through an LSI HBA to it to control the disks.

**Comment by tlehman** [comment]
As long as it shows up on the PCI bus with an address, we should be able to pass it through to the VM

**Comment by mircea-pavel-anton** [comment]
I see, that's great news! Do we have an ETA on this feature?

**Comment by tlehman** [comment]
Yeah, ETA is a month, aiming for releasing with v1.1.0

**Comment by mircea-pavel-anton** [comment]
That's Great! I am definitely looking forward to that!

**Comment by tlehman** [comment]
To clarify, the features targeted for 1.1.0 are all PCI-Passthrough based:

- SR-IOV in guest cluster mode (done by setting up SR-IOV in a VM after enabling PCI Passthrough on a network device(s) and assigning it/them to the VMs implementing the guest cluster)
- GPU passthrough by installing the drivers in the VM after enabling PCI Passthrough on a GPU device and assigning it to a VM

**Comment by tlehman** [comment]
Changed to remove the SR-IOV functionality. That is going to be added in a future release. This release (1.1.0 is going to ship PCI Passthrough only)

**Comment by tlehman** [comment]
PCI Passthrough will enable GPU support in VMs. Closing #1029 as a duplicate of this issue.

**Comment by tlehman** [comment]
Notes for release 1.1.0-rc1

The pcidevices controller should work for:
1. Detecting PCI Devices 
2. Creating PCI Device Claims (preparing the devices for Passthrough)
3. Attaching PCI Devices to VMs
4. Booting the VMs

The only thing that isn't done yet is the allow-list in the KubeVirt config. To get around this, you can edit the kubevirt CR manually and add in this YAML, generated with the following commands:

Generate YAML to allow-list all PCI devices for passthrough:
```
kubectl get pcidevices -o yaml | \                     git(816c…):kubevirt-cr k8s⎈ saturn 🪐
      yq '.items[].status |
          { "resourceName": .resourceName,
            "pciVendorSelector": .vendorId + ":" + .deviceId,
            "externalResourceProvider": false }'
```

Then edit `kubectl edit -n harvester-system kubevirts kubevirt`

under `spec.configuration` add: 

```
permittedHostDevices:
    pciHostDevices:
    - pciVendorSelector: "10DE:1EB8"
      resourceName: "nvidia.com/TU104GL_Tesla_T4"
      externalResourceProvider: true
    - pciVendorSelector: "8086:6F54"
      resourceName: "intel.com/qat"
```

With the results of the above `kubectl get` command. This will allow-list all the PCI devices for passthrough.

**Comment by tlehman** [comment]
Also, the UI changes for this PCI Passthrough feature are here: https://github.com/harvester/dashboard/pull/420

To test, if 420 is not merged yet, you can check this out and run `API=$MGMT_IP yarn dev` and using the URL that is printed on your screen. That will expose the Advanced -> PCI Devices menu item. From there you should be able to create PCI Device Claims and then in the VM UI you should be able to attach them to the VM. Slack me if you run into trouble

Call me +1 (425) 314-1497 if no Slack response. let's get this tested and shipped

**Comment by noahgildersleeve** [comment]
The dashboard PR was changed to this one https://github.com/harvester/dashboard/pull/432

**Comment by mgerboc** [comment]
Hello, when I try to enable nvidia device, the enabling is always in progress and I can see some errors in pci device controller pod. See attachment. 
[harvester-harvester-pcidevices-controller-vf2cf_agent.log](https://github.com/harvester/harvester/files/9733417/harvester-harvester-pcidevices-controller-vf2cf_agent.log)

**Comment by tlehman** [comment]
I have a PR that fixes this: https://github.com/harvester/pcidevices/pull/21

**Comment by noahgildersleeve** [comment]
Tested with custom ISO including some of the PCI fixes `master-937f333c-head`. 

I tried to follow the steps [here](https://github.com/harvester/tests/issues/529)

> ## Case 2 (GPU passthrough)
> 1. Create a harvester cluster in bare metal mode. Ensure one of the nodes has a GPU separate from the management NIC
> 2. Go to the management interface of the new cluster
> 3. Go to Advanced -> PCI Devices
> 4. Check the box representing the GPU device (identify it by the Description or the VendorId/DeviceId combination)
> 5. Click Enable Passthrough
> 6. When the GPU device is in an Enabled state, create a VM
> 7. After creating the VM, edit the Config
> 8. In the "PCI Devices" section, click the "Available PCI Devices" dropdown
> 9. Select the GPU device that has been enabled for passthrough
> 10. Click Save
> 11. Start the VM
> 12. Once the VM is booted, run `lspci` at the command line (make sure the VM has the `pciutils` package) and verify that the GPU device shows up
> 13. Install the driver for your GPU device
>      1. if the device is from NVIDIA: `sudo apt install nvidia-driver-510-server nvidia-utils-510 nvidia-cuda-toolkit`
>          1. Check out https://github.com/nvidia/cuda-samples
>          2. `cd cuda-samples/Samples/3_CUDA_Features/dmmaTensorCoreGemm`
>          3. `make` (if you run into problems with your cuda libraries, ping @tlehman )
>          4. Run `./dmmaTensorCoreGemm` and verify that the program completed correctly
>      2. if the device is from AMD/ATI, install and use the `aticonfig` command to inspect the device

Got errors when trying to start the VM after attaching the PCI device
![Greenshot 2022-10-13 11 53 55](https://user-images.githubusercontent.com/83787952/195695818-716ab2ce-6ae2-488b-b850-58ef49c6df72.png)

**Comment by noahgildersleeve** [comment]
I checked with @tlehman and there is apparently a fix for this in the most recent build of harvester/pcidevices. He's currently putting in a PR for it, but it will require a new build before testing. We might be able to build the ISO locally and test it.

**Comment by noahgildersleeve** [comment]
Tested in Harvester master-7b698413-head. I updated the test cases in https://github.com/harvester/tests/issues/529. Verified as working.

> ## Pre-requisite Enable PCI devices
> 1. Create a harvester cluster in bare metal mode. Ensure one of the nodes has NIC separate from the management NIC
> 2. Go to the management interface of the new cluster
> 3. Go to Advanced -> PCI Devices
> 4. Validate that the PCI devices aren't enabled
> 5. Click the link to enable PCI devices
> 6. Enable PCI devices in the linked addon page
> 7. Wait for the status to change to Deploy successful
> 8. Navigate to the PCI devices page
> 9. Validate that the PCI devices page is populated/populating with PCI devices
> 
> ## Case 1 (PCI NIC passthrough)
> 1. Create a harvester cluster in bare metal mode. Ensure one of the nodes has NIC separate from the management NIC
> 2. Go to the management interface of the new cluster
> 3. Go to Advanced -> PCI Devices
> 4. Check the box representing the PCI NIC device (identify it by the Description or the VendorId/DeviceId combination)
> 5. Click Enable Passthrough
> 6. When the NIC device is in an Enabled state, create a VM
> 7. After creating the VM, edit the Config
> 8. In the "PCI Devices" section, click the "Available PCI Devices" dropdown
> 9. Select the PCI NIC device that has been enabled for passthrough
> 10. Click Save
> 11. Start the VM
> 12. Once the VM is booted, run `lspci` at the command line (make sure the VM has the `pciutils` package) and verify that the PCI NIC device shows up
> 13. (Optional) Install the driver for your PCI NIC device (if it hasn't been autoloaded)
> 
> ### Case 1 dependencies:
> * PCI NIC separate from management network
> * Enable PCI devices
> 
> ## Case 2 (GPU passthrough)
> ### Case 2-1 Add GPU
> 1. Create a harvester cluster in bare metal mode. Ensure one of the nodes has a GPU separate from the management NIC
> 2. Go to the management interface of the new cluster
> 3. Go to Advanced -> PCI Devices
> 4. Check the box representing the GPU device (identify it by the Description or the VendorId/DeviceId combination)
> 5. Click Enable Passthrough
> 6. When the GPU device is in an Enabled state, create a VM
> 7. After creating the VM, edit the Config
> 8. In the "PCI Devices" section, click the "Available PCI Devices" dropdown
> 9. Select the GPU device that has been enabled for passthrough
> 10. Click Save
> 11. Start the VM
> 12. Once the VM is booted, run `lspci` at the command line (make sure the VM has the `pciutils` package) and verify that the GPU device shows up
> 13. Install the driver for your GPU device
>     
>     1. if the device is from NVIDIA: (this is for ubuntu, but the opensuse installation instructions are [here](https://docs.nvidia.com/cuda/cuda-installation-guide-linux/index.html#suse-installation))
>        ```
>        wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2004/x86_64/cuda-keyring_1.0-1_all.deb
>        sudo dpkg -i cuda-keyring_1.0-1_all.deb
>        sudo apt-get update
>        sudo apt-get -y install cuda
>        ```
>        
>        
>            
>              
>            
>        
>              
>            
>        
>            
>          
>        
>        1. Check out https://github.com/nvidia/cuda-samples
>           `git clone https://github.com/nvidia/cuda-samples`
>        2. `cd cuda-samples/Samples/3_CUDA_Features/dmmaTensorCoreGemm`
>        3. `make` (if you run into problems with your cuda libraries, ping @tlehman )
>        4. Run `./dmmaTensorCoreGemm` and verify that the program completed correctly
>     2. if the device is from AMD/ATI, install and use the `aticonfig` command to inspect the device
> 
> ### Case 2-2 Negative add GPU
> * **Pre-requisite**: the GPU should already be assigned to another VM
> 
> 1. On the VM that doesn't have the GPU assigned edit the VM
> 2. Open up the `PCI Devices` Section
> 3. Verify that you can't add the already assigned GPU to the VM. It should be greyed out.
> 
> ### Case 2-3 Remove GPU
> 1. Edit the VM where the GPU is assigned
> 2. In the "PCI Devices" section, Clear the available devices
> 3. Click save
>    
>    * If the VM is on then you will be prompted to reboot.
> 4. Validate that the GPU is removed by using `lspci` and trying to run `dmmaTensorCoreGemm` if the GPU supported CUDA
> 5. Open up another VM and verify that the GPU is listed as available in the `PCI Devices` section
